src/main/scala/au/com/agiledigital/toolform/tasks/generate/docker/GenerateDockerComposeV3.scala

```diff
@@ -0,0 +1,175 @@
+        } yield ()
+        _ <- resetIndent(i)
+      } yield ()
+
+      val context = WriterContext(writer)
+      writeFile.eval(context)
+
+      Right("Completed Successfully")
+    } finally {
+      writer.close()
+    }
+  }
+
+  // Sections
+
+  private def writeEdges(projectId: String, subEdgeDefs: List[SubEdgeDef]) = State[WriterContext, Unit] { context =>
+    (subEdgeDefs
+       .traverseU(subEdge => writeSubEdge(projectId, subEdge))
+       .exec(context),
+     ())
+  }
+
+  private def writeComponents(projectId: String, components: List[Component]) = State[WriterContext, Unit] { context =>
+    (components
+       .traverseU(component => writeComponent(projectId, component))
+       .exec(context),
+     ())
+  }
+
+  private def writeResources(resources: List[Resource]) = State[WriterContext, Unit] { context =>
+    (resources
+       .traverseU(resource => writeResource(resource))
+       .exec(context),
+     ())
+  }
+
+  private def writeSubEdge(projectId: String, subEdgeDef: SubEdgeDef) = {
+    val serviceName = subEdgeServiceName(projectId, subEdgeDef)
+    val imageName   = subEdgeImageName(projectId, subEdgeDef)
+    for {
+      _ <- write(s"$serviceName:")
+      i <- indent()
+      _ <- for {
+        _ <- write(s"image: $imageName")
+        _ <- write(s"restart: always")
+        _ <- write(s"ports:")
+        _ <- write(subEdgePortDefinition(subEdgeDef))
+      } yield ()
+      _ <- resetIndent(i)
+    } yield ()
+  }
+
+  private def writeComponent(projectId: String, component: Component) = {
+    val serviceName = componentServiceName(component)
+    val imageName   = componentImageName(projectId, component)
+    for {
+      _ <- write(s"$serviceName:")
+      i <- indent()
+      _ <- for {
+        _ <- write(s"image: $imageName")
+        _ <- write(s"restart: always")
+        _ <- writeComponentLabels(component)
+        _ <- writeEnvironmentVariables(component)
+        _ <- writePorts(component)
+      } yield ()
+      _ <- resetIndent(i)
+    } yield ()
+  }
+
+  private def writeComponentLabels(component: Component) =
+    for {
+      _ <- write("labels:")
+      i <- indent()
+      _ <- for {
+
+        _ <- write(s"source.path: \042${component.path}\042")
+        _ <- write("project.artefact: \"true\"")
+      } yield ()
+      _ <- resetIndent(i)
+    } yield ()
+
+  private def writeResource(resource: Resource) = {
+    val serviceName = resourceServiceName(resource)
+    val imageName   = resourceImageName(resource)
+    for {
+      _ <- write(s"$serviceName:")
+      i <- indent()
+      _ <- for {
+
+        _ <- write(s"image: $imageName")
+        _ <- write(s"restart: always")
+        _ <- writeEnvironmentVariables(resource)
+        _ <- writePorts(resource)
+      } yield ()
+      _ <- resetIndent(i)
+    } yield ()
+  }
+
+  private def writeEnvironmentVariables(service: Service) = service.environment match {
+    case Some(environmentEntries) =>
```

Comment:
`environment` should be a `Seq[...]` not `Option[Seq[...]]`, I think that there were some issues making this true earlier. Should fix now so that it isn't bleeding into the generators.

Reply:
It seems like the parser throws an exception if the key isn't defined. I think we would have to make it mandatory to define empty environment/port entries.

Reply:
Oh never mind you can specify default values on case classes

Reply:
Fixed by specifying a default value for environment and exposed ports in the case class.

This has now be simplified to an `if` statement checking if the list/map is empty or not.